pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -364,6 +364,16 @@ func (c *Controller) syncService(key string) error {
 		return err
 	}
 
+	// Drop EndpointSlices that have been marked for deletion to prevent the controller from getting stuck.
+	n := 0
+	for _, endpointSlice := range endpointSlices {
+		if endpointSlice.DeletionTimestamp == nil {
+			endpointSlices[n] = endpointSlice
+			n++
+		}
+	}
+	endpointSlices = endpointSlices[:n]
```

Comment:
I think this would be better as a separate function like `trimTerminatingSlices()` that could also have a simple unit test covering it.